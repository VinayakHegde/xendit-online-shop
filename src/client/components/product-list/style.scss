.ProductList {
  display: grid;
  gap: 20px;
  grid-template-rows: auto  1fr auto;
  &__wrapper {
    list-style: none;
    display: grid;
    gap: 20px;
    padding: 0;
  }
  &__item {
    display: grid;
    gap: 20px;
    grid-template-columns: 100px 200px 1fr;
    border-bottom: 1px solid gainsboro;
    padding-bottom: 20px;
    &__image{
      width: 100px;
    }
    &__info {
      &:nth-child(2) {
        display: grid;
        gap: 5px;
        grid-template-rows: auto 1fr auto;
      }
      &__description {
        text-align: justify;
      }
    }
  }
  &__cart__action {
    display: grid;
    grid-template-columns: 1fr auto;
    align-items: center;
    text-align: right;
    gap: 20px;
}
}
