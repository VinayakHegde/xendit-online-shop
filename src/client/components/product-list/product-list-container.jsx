import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { getProducts } from "../../store/product-slice";
import { ProductList } from "./product-list";
import './style.scss';

export const ProductListContainer = ({ title = "Product list", isFromCart = false }) => {
  const products = useSelector(state => {
    if (isFromCart) {
      return state.cart.data || [];
    }
    return state.products.data || []
  });
  const dispatch = useDispatch();
  useEffect(() => {
    if (!isFromCart) {
      dispatch(getProducts());
    }
  }, [isFromCart]);
  return (
    <ProductList products={products} title={title} isFromCart={isFromCart} />
  );
}
