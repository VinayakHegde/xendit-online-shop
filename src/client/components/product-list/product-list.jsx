import React from "react";
import { AddToCart } from "../add-to-cart";
import { ProceedTo } from "../proceed-to";
import { ProductImage, ProductStock, ProductDescription, ProductName, ProductPrice } from "../product";
import { ProductQuantity } from "../product/product-quantity";
import './style.scss';

export const ProductList = ({
  products = [],
  title = "Product list",
  isFromCart = false
}) => (
  <section className="ProductList">
    <div>
      <h3>{isFromCart && !products.length ? 'Cart is empty' : title}</h3>
    </div>
    {(products.length > 0) && (
      <ul className="ProductList__wrapper">
        {products.map(product => (
          <li key={product.id} className="ProductList__item">
            <div className="ProductList__item__info">
              <ProductImage src={product.image} id={product.id} classPrefix="ProductList__item" />
              <ProductStock quantity={product.quantity} noRender={isFromCart} />
            </div>
            <div className="ProductList__item__info">
              <ProductName name={product.name} id={product.id} classPrefix="ProductList__item" />
              <ProductPrice price={product.price} classPrefix="ProductList__item" />
              <AddToCart product={product} classPrefix="ProductList__item" noRender={isFromCart} />
              <ProductQuantity quantity={product.quantity} noRender={!isFromCart} />
            </div>
            <ProductPrice price={product.price * product.quantity} caption="Total Price" classPrefix="ProductList__item" noRender={!isFromCart} />
            <ProductDescription descr={product.description} classPrefix="ProductList__item" noRender={isFromCart} />
          </li>
        ))}
      </ul>
    )}
    {isFromCart && (
      <div className="ProductList__cart__action">
        <ProceedTo />
      </div>
    )}
  </section>
);
