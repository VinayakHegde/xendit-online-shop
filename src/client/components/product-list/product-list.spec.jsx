import React from "react";
import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import * as ReactRedux from 'react-redux';
import { fromString } from "uuidv4";
import { ProductList } from "./product-list";

afterAll(() => {
  jest.clearAllMocks();
});
const spyOnUseDispatch = (mock, expected) =>
  jest.spyOn(ReactRedux, 'useDispatch').mockImplementation(() => mock.mockResolvedValue(expected));

const getProps = (props) => {
  return {
    products: [{
      id: fromString('Sunglass'),
      name: 'Sunglass',
      price: 9299,
      image: 'sunglasses.png',
      quantity: 10,
      description: 'Sunglasses or sun glasses (informally called shades or sunnies) are a form of protective eyewear designed primarily to prevent bright sunlight and high-energy visible light from damaging or discomforting the eyes. They can sometimes also function as a visual aid, as variously termed spectacles or glasses exist, featuring lenses that are colored, polarized or darkened. In the early 20th century, they were also known as sun cheaters'
    }, {
      id: fromString('Headphone'),
      name: 'Headphone',
      price: 5049,
      image: 'headphones.png',
      quantity: 10,
      description: 'Headphones are a pair of small loudspeaker drivers worn on or around the head over a user\'s ears. They are electroacoustic transducers, which convert an electrical signal to a corresponding sound. Headphones let a single user listen to an audio source privately, in contrast to a loudspeaker, which emits sound into the open air for anyone nearby to hear. Headphones are also known as earspeakers, earphones, colloquially, cans.',
    }, {
      id: fromString('Chair'),
      name: 'Chair',
      price: 8000,
      image: 'chair.png',
      quantity: 10,
      description: 'Chairs vary in design. An armchair has armrests fixed to the seat, a recliner is upholstered and under its seat is a mechanism that allows one to lower the chair\'s back and raise into place a fold-out footrest, a rocking chair has legs fixed to two long curved slats and a wheelchair has wheels fixed to an axis under the seat.',
    }, {
      id: fromString('Flashlight'),
      name: 'Flashlight',
      price: 2099,
      image: 'flashlight.png',
      quantity: 3,
      description: 'A flashlight (US) or torch (UK) is a portable hand-held electric light. The source of the light often used to be an incandescent light bulb (lamp) but has been gradually replaced by light-emitting diodes (LEDs) since the mid-2000s. A typical flashlight consists of the light source mounted in a reflector, a transparent cover (sometimes combined with a lens) to protect the light source and reflector, a battery, and a switch. These are supported and protected by a case.',
    }, {
      id: fromString('Phone'),
      name: 'Phone',
      price: 40000,
      image: 'phone.png',
      quantity: props.isFromCart ? 10 : 0,
      description: 'A mobile phone, cellular phone, cell phone, cellphone, handphone, or hand phone, sometimes shortened to simply mobile, cell or just phone, is a portable telephone that can make and receive calls over a radio frequency link while the user is moving within a telephone service area. ',
    }],
    title: "Product list",
    isFromCart: false,
    ...props
  }
}
describe.each([
  ["Product list", getProps({})],
  ["Cart", getProps({
    title: "Cart",
    isFromCart: true
  })],
  ["Empty Cart", getProps({
    products: [],
    isFromCart: true
  })]
])('ProductList %s', (a, props) => {
  it('should render without crash', () => {
    jest.spyOn(ReactRedux, 'useSelector').mockImplementation(() => { });
    spyOnUseDispatch(jest.fn(), {});
    const { container } = render(<BrowserRouter><ProductList {...props} /></BrowserRouter>)
    expect(container).toMatchSnapshot();
  });
});
