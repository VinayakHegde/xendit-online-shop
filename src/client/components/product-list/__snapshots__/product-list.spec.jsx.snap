// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductList Cart should render without crash 1`] = `
<div>
  <section
    class="ProductList"
  >
    <div>
      <h3>
        Cart
      </h3>
    </div>
    <ul
      class="ProductList__wrapper"
    >
      <li
        class="ProductList__item"
      >
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/49624fbd-6937-5b88-810d-0dc5af14f2c0"
          >
            <img
              alt=""
              class="ProductList__item__image"
              src="/images/sunglasses.png"
            />
          </a>
        </div>
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/49624fbd-6937-5b88-810d-0dc5af14f2c0"
          >
            <strong
              class="ProductList__item__name"
            >
              Sunglass
            </strong>
          </a>
          <div
            class="ProductList__item__info__price"
          >
            <span
              class="ProductList__item__label"
            >
              Price
              : 
            </span>
            <span
              class="ProductList__item__price"
            >
              9299
            </span>
          </div>
          <div
            class="ProductQuantity__info__quantity"
          >
            <span
              class="ProductQuantity__label"
            >
              Quantity: 
            </span>
            <span
              class="ProductQuantity__quantity"
            >
              10
            </span>
          </div>
        </div>
        <div
          class="ProductList__item__info__price"
        >
          <span
            class="ProductList__item__label"
          >
            Total Price
            : 
          </span>
          <span
            class="ProductList__item__price"
          >
            92990
          </span>
        </div>
      </li>
      <li
        class="ProductList__item"
      >
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/0d444f8b-c35d-5812-bfc8-8be6fb7252b5"
          >
            <img
              alt=""
              class="ProductList__item__image"
              src="/images/headphones.png"
            />
          </a>
        </div>
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/0d444f8b-c35d-5812-bfc8-8be6fb7252b5"
          >
            <strong
              class="ProductList__item__name"
            >
              Headphone
            </strong>
          </a>
          <div
            class="ProductList__item__info__price"
          >
            <span
              class="ProductList__item__label"
            >
              Price
              : 
            </span>
            <span
              class="ProductList__item__price"
            >
              5049
            </span>
          </div>
          <div
            class="ProductQuantity__info__quantity"
          >
            <span
              class="ProductQuantity__label"
            >
              Quantity: 
            </span>
            <span
              class="ProductQuantity__quantity"
            >
              10
            </span>
          </div>
        </div>
        <div
          class="ProductList__item__info__price"
        >
          <span
            class="ProductList__item__label"
          >
            Total Price
            : 
          </span>
          <span
            class="ProductList__item__price"
          >
            50490
          </span>
        </div>
      </li>
      <li
        class="ProductList__item"
      >
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/b483939a-6fe9-5c47-acc8-ec9741e8a8c7"
          >
            <img
              alt=""
              class="ProductList__item__image"
              src="/images/chair.png"
            />
          </a>
        </div>
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/b483939a-6fe9-5c47-acc8-ec9741e8a8c7"
          >
            <strong
              class="ProductList__item__name"
            >
              Chair
            </strong>
          </a>
          <div
            class="ProductList__item__info__price"
          >
            <span
              class="ProductList__item__label"
            >
              Price
              : 
            </span>
            <span
              class="ProductList__item__price"
            >
              8000
            </span>
          </div>
          <div
            class="ProductQuantity__info__quantity"
          >
            <span
              class="ProductQuantity__label"
            >
              Quantity: 
            </span>
            <span
              class="ProductQuantity__quantity"
            >
              10
            </span>
          </div>
        </div>
        <div
          class="ProductList__item__info__price"
        >
          <span
            class="ProductList__item__label"
          >
            Total Price
            : 
          </span>
          <span
            class="ProductList__item__price"
          >
            80000
          </span>
        </div>
      </li>
      <li
        class="ProductList__item"
      >
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/5b332c6b-ae5f-58d1-a383-681a993034a3"
          >
            <img
              alt=""
              class="ProductList__item__image"
              src="/images/flashlight.png"
            />
          </a>
        </div>
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/5b332c6b-ae5f-58d1-a383-681a993034a3"
          >
            <strong
              class="ProductList__item__name"
            >
              Flashlight
            </strong>
          </a>
          <div
            class="ProductList__item__info__price"
          >
            <span
              class="ProductList__item__label"
            >
              Price
              : 
            </span>
            <span
              class="ProductList__item__price"
            >
              2099
            </span>
          </div>
          <div
            class="ProductQuantity__info__quantity"
          >
            <span
              class="ProductQuantity__label"
            >
              Quantity: 
            </span>
            <span
              class="ProductQuantity__quantity"
            >
              3
            </span>
          </div>
        </div>
        <div
          class="ProductList__item__info__price"
        >
          <span
            class="ProductList__item__label"
          >
            Total Price
            : 
          </span>
          <span
            class="ProductList__item__price"
          >
            6297
          </span>
        </div>
      </li>
      <li
        class="ProductList__item"
      >
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/a2b24884-dd89-505c-81e2-afa0baa8a3a2"
          >
            <img
              alt=""
              class="ProductList__item__image"
              src="/images/phone.png"
            />
          </a>
        </div>
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/a2b24884-dd89-505c-81e2-afa0baa8a3a2"
          >
            <strong
              class="ProductList__item__name"
            >
              Phone
            </strong>
          </a>
          <div
            class="ProductList__item__info__price"
          >
            <span
              class="ProductList__item__label"
            >
              Price
              : 
            </span>
            <span
              class="ProductList__item__price"
            >
              40000
            </span>
          </div>
          <div
            class="ProductQuantity__info__quantity"
          >
            <span
              class="ProductQuantity__label"
            >
              Quantity: 
            </span>
            <span
              class="ProductQuantity__quantity"
            >
              10
            </span>
          </div>
        </div>
        <div
          class="ProductList__item__info__price"
        >
          <span
            class="ProductList__item__label"
          >
            Total Price
            : 
          </span>
          <span
            class="ProductList__item__price"
          >
            400000
          </span>
        </div>
      </li>
    </ul>
    <div
      class="ProductList__cart__action"
    >
      <a
        href="/"
      >
        <button
          class="Button undefined"
          title="Start Shopping"
          type="button"
        >
          Start Shopping
        </button>
      </a>
    </div>
  </section>
</div>
`;

exports[`ProductList Empty Cart should render without crash 1`] = `
<div>
  <section
    class="ProductList"
  >
    <div>
      <h3>
        Cart is empty
      </h3>
    </div>
    <div
      class="ProductList__cart__action"
    >
      <a
        href="/"
      >
        <button
          class="Button undefined"
          title="Start Shopping"
          type="button"
        >
          Start Shopping
        </button>
      </a>
    </div>
  </section>
</div>
`;

exports[`ProductList Product list should render without crash 1`] = `
<div>
  <section
    class="ProductList"
  >
    <div>
      <h3>
        Product list
      </h3>
    </div>
    <ul
      class="ProductList__wrapper"
    >
      <li
        class="ProductList__item"
      >
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/49624fbd-6937-5b88-810d-0dc5af14f2c0"
          >
            <img
              alt=""
              class="ProductList__item__image"
              src="/images/sunglasses.png"
            />
          </a>
        </div>
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/49624fbd-6937-5b88-810d-0dc5af14f2c0"
          >
            <strong
              class="ProductList__item__name"
            >
              Sunglass
            </strong>
          </a>
          <div
            class="ProductList__item__info__price"
          >
            <span
              class="ProductList__item__label"
            >
              Price
              : 
            </span>
            <span
              class="ProductList__item__price"
            >
              9299
            </span>
          </div>
          <button
            class="Button AddToCart ProductList__item__button"
            title="Add to cart"
            type="button"
          >
            Add to cart
          </button>
        </div>
        <div
          class="ProductList__item__info__description"
        >
          Sunglasses or sun glasses (informally called shades or sunnies) are a form of protective eyewear designed primarily to prevent bright sunlight and high-energy visible light from damaging or discomforting the eyes. They can sometimes also function as a visual aid, as variously termed spectacles or glasses exist, featuring lenses that are colored, polarized or darkened. In the early 20th century, they were also known as sun cheaters
        </div>
      </li>
      <li
        class="ProductList__item"
      >
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/0d444f8b-c35d-5812-bfc8-8be6fb7252b5"
          >
            <img
              alt=""
              class="ProductList__item__image"
              src="/images/headphones.png"
            />
          </a>
        </div>
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/0d444f8b-c35d-5812-bfc8-8be6fb7252b5"
          >
            <strong
              class="ProductList__item__name"
            >
              Headphone
            </strong>
          </a>
          <div
            class="ProductList__item__info__price"
          >
            <span
              class="ProductList__item__label"
            >
              Price
              : 
            </span>
            <span
              class="ProductList__item__price"
            >
              5049
            </span>
          </div>
          <button
            class="Button AddToCart ProductList__item__button"
            title="Add to cart"
            type="button"
          >
            Add to cart
          </button>
        </div>
        <div
          class="ProductList__item__info__description"
        >
          Headphones are a pair of small loudspeaker drivers worn on or around the head over a user's ears. They are electroacoustic transducers, which convert an electrical signal to a corresponding sound. Headphones let a single user listen to an audio source privately, in contrast to a loudspeaker, which emits sound into the open air for anyone nearby to hear. Headphones are also known as earspeakers, earphones, colloquially, cans.
        </div>
      </li>
      <li
        class="ProductList__item"
      >
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/b483939a-6fe9-5c47-acc8-ec9741e8a8c7"
          >
            <img
              alt=""
              class="ProductList__item__image"
              src="/images/chair.png"
            />
          </a>
        </div>
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/b483939a-6fe9-5c47-acc8-ec9741e8a8c7"
          >
            <strong
              class="ProductList__item__name"
            >
              Chair
            </strong>
          </a>
          <div
            class="ProductList__item__info__price"
          >
            <span
              class="ProductList__item__label"
            >
              Price
              : 
            </span>
            <span
              class="ProductList__item__price"
            >
              8000
            </span>
          </div>
          <button
            class="Button AddToCart ProductList__item__button"
            title="Add to cart"
            type="button"
          >
            Add to cart
          </button>
        </div>
        <div
          class="ProductList__item__info__description"
        >
          Chairs vary in design. An armchair has armrests fixed to the seat, a recliner is upholstered and under its seat is a mechanism that allows one to lower the chair's back and raise into place a fold-out footrest, a rocking chair has legs fixed to two long curved slats and a wheelchair has wheels fixed to an axis under the seat.
        </div>
      </li>
      <li
        class="ProductList__item"
      >
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/5b332c6b-ae5f-58d1-a383-681a993034a3"
          >
            <img
              alt=""
              class="ProductList__item__image"
              src="/images/flashlight.png"
            />
          </a>
          <div
            class="ProductStock__low"
          >
            <small>
              only 
              3
               left
            </small>
          </div>
        </div>
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/5b332c6b-ae5f-58d1-a383-681a993034a3"
          >
            <strong
              class="ProductList__item__name"
            >
              Flashlight
            </strong>
          </a>
          <div
            class="ProductList__item__info__price"
          >
            <span
              class="ProductList__item__label"
            >
              Price
              : 
            </span>
            <span
              class="ProductList__item__price"
            >
              2099
            </span>
          </div>
          <button
            class="Button AddToCart ProductList__item__button"
            title="Add to cart"
            type="button"
          >
            Add to cart
          </button>
        </div>
        <div
          class="ProductList__item__info__description"
        >
          A flashlight (US) or torch (UK) is a portable hand-held electric light. The source of the light often used to be an incandescent light bulb (lamp) but has been gradually replaced by light-emitting diodes (LEDs) since the mid-2000s. A typical flashlight consists of the light source mounted in a reflector, a transparent cover (sometimes combined with a lens) to protect the light source and reflector, a battery, and a switch. These are supported and protected by a case.
        </div>
      </li>
      <li
        class="ProductList__item"
      >
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/a2b24884-dd89-505c-81e2-afa0baa8a3a2"
          >
            <img
              alt=""
              class="ProductList__item__image"
              src="/images/phone.png"
            />
          </a>
          <div
            class="ProductStock__outofstock"
          >
            <small>
              out of stock
            </small>
          </div>
        </div>
        <div
          class="ProductList__item__info"
        >
          <a
            class="ProductList__item__link"
            href="/product/a2b24884-dd89-505c-81e2-afa0baa8a3a2"
          >
            <strong
              class="ProductList__item__name"
            >
              Phone
            </strong>
          </a>
          <div
            class="ProductList__item__info__price"
          >
            <span
              class="ProductList__item__label"
            >
              Price
              : 
            </span>
            <span
              class="ProductList__item__price"
            >
              40000
            </span>
          </div>
          <button
            class="Button AddToCart ProductList__item__button"
            disabled=""
            title="Add to cart"
            type="button"
          >
            Add to cart
          </button>
        </div>
        <div
          class="ProductList__item__info__description"
        >
          A mobile phone, cellular phone, cell phone, cellphone, handphone, or hand phone, sometimes shortened to simply mobile, cell or just phone, is a portable telephone that can make and receive calls over a radio frequency link while the user is moving within a telephone service area. 
        </div>
      </li>
    </ul>
  </section>
</div>
`;
